File: frontend/src/types/trace.ts
export type RuntimeTraceEvent = {
  id: string;
  type: string;
  timestamp: string;
  step: number;
  file_path: string;
  line_number: number | null;
  scope: Record<string, unknown>;
  payload: Record<string, unknown>;
  visual: Record<string, unknown>;
  validation: Record<string, unknown> | null;
};

export type RuntimeTraceError = {
  file_path: string | null;
  message: string;
  line_number: number | null;
};

export type RuntimeTraceResult = {
  events: RuntimeTraceEvent[];
  stdout: string;
  stderr: string;
  errors: RuntimeTraceError[];
};

File: backend/app/models/trace.py
from pydantic import BaseModel, Field

from app.models.analysis import AnalysisError, AnalysisFile


class RuntimeTraceRequest(BaseModel):
    files: list[AnalysisFile] = Field(default_factory=list)
    entry_file: str = Field(..., min_length=1)


class RuntimeTraceEvent(BaseModel):
    id: str
    type: str
    timestamp: str
    step: int
    file_path: str
    line_number: int | None
    scope: dict
    payload: dict = Field(default_factory=dict)
    visual: dict = Field(default_factory=dict)
    validation: dict | None = None


class RuntimeTraceResponse(BaseModel):
    events: list[RuntimeTraceEvent]
    stdout: str
    stderr: str
    errors: list[AnalysisError]
